syntax = "proto3";

package tapioca.core;

message ManifestProto {
  repeated ManifestBlockProto blocks = 1;
  repeated ManifestItemProto items = 2;
  // Required: the maximum size of blocks within the build in bytes.
  uint64 max_block_size = 3;
}

// A manifest descriptor for a block within a build
message ManifestBlockProto {
  // The hash of the block
  bytes hash = 1;
  // Optional: The size of the block. Only present when the size of the block is
  // not the max_block_size of the manifest.
  uint64 size = 2;
}

// A manifest descriptor for a item (a file or directory) within a build
message ManifestItemProto {
  string name = 1;

  oneof descriptor {
    File file = 2;
    Directory directory = 3;
  }

  message File {
    bytes hash = 1;
    uint64 size = 2;
    // The sequence of blocks that consitute the file. The ID corresponds to the
    // index of the block in the manifest.
    repeated uint64 block_ids = 3;
  }

  message Directory {
    repeated ManifestItemProto children = 1;
  }
}
